### ISSUE TITLE
artifact expiration is not verified at task submission time

### ISSUE THREAD

**Issue by bhearsum** [taskcluster/taskcluster#7416]
[This task](https://stage.taskcluster.nonprod.cloudops.mozgcp.net/tasks/Ujo9jB_GS52cICfFAVTL1Q/definition) was successfully submitted and was scheduled to run. Notable details of its definition include:
```
created: '2024-12-12T19:27:20.756Z'
deadline: '2024-12-13T19:27:20.756Z'
expires: '2025-03-12T19:27:20.756Z'
payload:
  artifacts:
    public/build:
      path: /builds/worker/artifacts
      type: directory
      expires: '2025-12-09T17:17:13.863Z'
```

It was not until it began to run that an error was thrown.

I assume this happens because artifacts are part of the payload, and the Queue doesn't validate that at all. This is still a bit surprising though, in that this is a docker-worker task, and from a distance, it seems like Taskcluster ought to be able to detect this earlier. (I would assume the same problem applies to a generic-worker task, but I haven't been able to verify that.) I don't know if it's possible or practical to fix this, but I wanted to file it in any case.

(It is, of course, fair to assume that the Queue could not possibly catch errors for tasks intended for other worker types, like scriptworker, at submission time.)

**Comment by lotas** [comment]
Thanks Ben for adding this.
I see that queue has a 15min tolerance for created artifacts (can be up to 15min expire into the past).
And true, queue doesn't know how to validate task payloads, as it is worker-dependent. 

This probably comes back to the problem that task queues/worker pools don't say which type of payload schema they support, and before you have at least one worker running there is hardly a way to assume which kind of payload this is.

Another issue I see here, even if you create a task with artifact expiry in the future, we can't guarantee that by the time task would be picked up by a worker it wouldn't be expired already. 

Technically, we can guess it's either docker or g-w, as they are the most popular and mostly used ones, and gracefully check the payload using their task schemas. 

I'm not sure what the best fix here is.

**Comment by bhearsum** [comment]
> Another issue I see here, even if you create a task with artifact expiry in the future, we can't guarantee that by the time task would be picked up by a worker it wouldn't be expired already.

That's a good point, yeah. Clearly even if _some_ effort was made here it could not possibly catch _all_ issues with expirations. Arguably, it would be good to catch timestamps in the past as early as possible, but I'm not sure it's worth the effort.

**Comment by lotas** [comment]
maybe we could extend worker pools with an optional payload validation schema? 
Might be easier then trying to guess which payload format is being used, and pools usually know what type of tasks/workers they have. 

The problem is that not all task queues have related worker pools in worker manager :) 

Something like `queue.forcePayloadSchema(taskQueueId, schema)` ? :)

**Comment by petemoore** [comment]
I think if a task is submitted with an artifact expiry in the past, it is probably a one-off task that somebody has hand crafted. I'm not sure there is a lot of value optimising for that case. Most tasks will be generated by automated tools, and won't exhibit this issue. Perhaps the root issue is that tasks take an unpleasantly long time to be claimed? Maybe fewer but larger worker pools could help reduce claim time, I'm not sure if we have a way to achieve that in FXCI? Ideally tasks would be claimed within seconds of them being submitted, and then I think it wouldn't matter too much if it was the Queue or the worker that validated the payload.

**Comment by petemoore** [comment]
Note, in the past we have talked about workers submitting their payload schema to the Queue for early validation of tasks, but that also would not have caught it, as json schema isn't expressive enough to determine whether a static timestamp is in the past.
